### PR TITLE
Implement communication task

### PR DESCRIPTION
**Pull request by adam-dej**
This PR implements the communication task.

Things left to do:

  - [x] Get DCRCP and DCL2 PRs merged and update libdeadcom submodule accordingly (https://github.com/fmfi-svt-deadlock/libdeadcom/pull/1 and https://github.com/fmfi-svt-deadlock/libdeadcom/pull/2)
  - [x] Fix transmit function returning as soon as bytes are copied to the output buffer, as opposed to transmitted
  - [x] Fix hardcoded length constants

### CODE REVIEW

**Comment by adam-dej** [comment]
This PR is no longer WIP, @lubo please take a look

File: src/tasks/comm-task/comm-task.c [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+/*===========================================================================*/
+/* Task-specific API                                                         */
+/*===========================================================================*/
+
+void dlTaskCommSendSysQueryResp(uint16_t rdrClass, uint16_t hwModel, uint16_t hwRev,
+                                char serial[DCRCP_SERIAL_MAX_LEN], uint8_t swVerMajor,
+                                uint8_t swVerMinor) {
+    DeadcomCRPM *m = chGuardedPoolAllocTimeout(&out_pool, TIME_INFINITE);
+    if (m == NULL) {
+        chSysHalt("null from inifinitely waiting guarded pool");
```
Comment (lubo) [inline src/tasks/comm-task/comm-task.c]: \* infinitely

File: src/tasks/comm-task/comm-task.c [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    m->data.sysQueryResponse.swVerMajor = swVerMajor;
+    m->data.sysQueryResponse.swVerMinor = swVerMinor;
+    memcpy(&(m->data.sysQueryResponse.serial), serial, DCRCP_SERIAL_MAX_LEN);
+    (void)chMBPost(&outbox, (msg_t)m, TIME_INFINITE);
+}
+
+void dlTaskCommSendRdrFailure(char *str) {
+    DeadcomCRPM *m = chGuardedPoolAllocTimeout(&out_pool, TIME_INFINITE);
+    if (m == NULL) {
+        chSysHalt("null from inifinitely waiting guarded pool");
```
Comment (lubo) [inline src/tasks/comm-task/comm-task.c]: \* infinitely

File: src/tasks/comm-task/comm-task.c [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    memset(m, 0, sizeof(DeadcomCRPM));
+    m->type = DCRCP_CRPM_RDR_FAILURE;
+    strncpy(m->data.rdrFailure, str, DCRCP_RDR_FAILURE_MSG_MAX_LEN);
+    (void)chMBPost(&outbox, (msg_t)m, TIME_INFINITE);
+}
+
+void dlTaskCommSendAM0GotUids(dl_picc_uid *uids, size_t uids_len) {
+    DeadcomCRPM *m = chGuardedPoolAllocTimeout(&out_pool, TIME_INFINITE);
+    if (m == NULL) {
+        chSysHalt("null from inifinitely waiting guarded pool");
```
Comment (lubo) [inline src/tasks/comm-task/comm-task.c]: \* infinitely

File: src/tasks/comm-task/comm-task.c [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        return;
+    }
+    memset(m, 0, sizeof(DeadcomCRPM));
+    m->type = DCRCP_CRPM_AM0_PICC_UID_OBTAINED;
+    m->data.authMethod0UuidObtained.len = uids_len;
+    size_t i;
+    for (i = 0; i < uids_len; i++) {
```
Comment (lubo) [inline src/tasks/comm-task/comm-task.c]: Can't you move definition of `i` into scope of this for?

File: src/tasks/comm-task/comm-task.c [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+            size_t out_size;
+            DCRCPStatus s = dcrcpEncode(m, data, DEADCOM_PAYLOAD_MAX_LEN, &out_size,
+                                        &cbor_out_context);
+            chDbgAssert(s == DCRCP_STATUS_OK, "encoding failed");
+            dcSendMessage(&dc_link, data, out_size);
+
+            chGuardedPoolFree(&out_pool, m);
+        }
+
+        //TODO generate heartbeat
```
Comment (lubo) [inline src/tasks/comm-task/comm-task.c:176]: Is this TODO harmless?

File: src/tasks/comm-task/comm-task.c [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+                    break;
+                default:
+                    // It is safe to ignore any other frame
+                    break;
+            }
+        }
+
+        heartbeat:
+        ;
+        // TODO generate heartbeat
```
Comment (lubo) [inline src/tasks/comm-task/comm-task.c:246]: Is this TODO harmless?

**Review by lubo — CHANGES_REQUESTED** [review]

File: src/tasks/comm-task/comm-task.c (hunk 5 above)
Comment (adam-dej) [reply src/tasks/comm-task/comm-task.c:176]: Yes, heartbeats will be implemented in Master Task, which will be in a different PR

File: src/tasks/comm-task/comm-task.c (hunk 6 above)
Comment (adam-dej) [reply src/tasks/comm-task/comm-task.c:246]: Yes, heartbeats will be implemented in Master Task, which will be in a different PR

File: src/tasks/comm-task/comm-task.c (hunk 1 above)
Comment (adam-dej) [reply src/tasks/comm-task/comm-task.c]: fixed

File: src/tasks/comm-task/comm-task.c (hunk 2 above)
Comment (adam-dej) [reply src/tasks/comm-task/comm-task.c]: fixed

File: src/tasks/comm-task/comm-task.c (hunk 3 above)
Comment (adam-dej) [reply src/tasks/comm-task/comm-task.c]: fixed

File: src/tasks/comm-task/comm-task.c (hunk 4 above)
Comment (adam-dej) [reply src/tasks/comm-task/comm-task.c]: moved

**Review by lubo — APPROVED** [review]

**Pull request merged**